core/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -188,6 +214,76 @@ public SimpleQueryBus(MessageMonitor<? super QueryMessage<?, ?>> messageMonitor,
     }
 
+    @SuppressWarnings("unchecked")
+    @Override
+    public <Q, I, U> Registration subscriptionQuery(SubscriptionQueryMessage<Q, I, U> query,
+                                                    UpdateHandler<I, U> updateHandler) {
+        MessageMonitor.MonitorCallback monitorCallback = messageMonitor.onMessageIngested(query);
+        SubscriptionQueryMessage<Q, I, U> interceptedQuery = intercept(query);
+        List<QuerySubscription> subs = subscriptions
+                .computeIfAbsent(interceptedQuery.getQueryName(), k -> new CopyOnWriteArrayList<>())
+                .stream()
+                .filter(subscription -> interceptedQuery.getResponseType().matches(subscription.getResponseType()))
+                .collect(Collectors.toList());
+        Registration registration = () -> true;
+        try {
+            if (subs.isEmpty()) {
+                throw new NoHandlerForQueryException(
+                        format("No handler found for %s with response type %s and update type %s",
+                               interceptedQuery.getQueryName(),
+                               interceptedQuery.getResponseType(),
+                               interceptedQuery.getUpdateResponseType()));
+            }
+            Iterator<QuerySubscription> subsIterator = subs.iterator();
+            boolean invocationSuccess = false;
+            SimpleQueryUpdateEmitter emitter = new SimpleQueryUpdateEmitter(updateHandler);
+            while (!invocationSuccess && subsIterator.hasNext()) {
+                try {
+                    QuerySubscription subscription = subsIterator.next();
+                    if (subscription instanceof SubscribableQuerySubscription) {
+                        DefaultUnitOfWork<QueryMessage<Q, I>> uow = DefaultUnitOfWork.startAndGet(
+                                interceptedQuery);
+                        if (interceptedQuery.getUpdateResponseType()
+                                            .matches(((SubscribableQuerySubscription) subscription).getUpdateType())) {
+
+                            registration = () -> {
+                                emitter.unsubscribeUpdateHandler();
+                                return true;
+                            };
+                            I initialResult = interceptAndInvoke(uow,
+                                                                 m -> ((SubscribableQuerySubscription) subscription)
+                                                                         .getSubscriptionQueryHandler()
+                                                                         .handle(m, emitter))
+                                    .getPayload();
+                            invocationSuccess = true;
+                            updateHandler.onInitialResult(initialResult);
+                        }
+                    } else {
+                        DefaultUnitOfWork<QueryMessage<Q, I>> uow = DefaultUnitOfWork.startAndGet(interceptedQuery);
```

Comment:
I think I'd prefer to have a descriptive, `private` function performing the work in the `else` block. Might make understanding the code somewhat clearer.